ambry-router/src/main/java/com.github.ambry.router/GetOperation.java

```diff
@@ -170,16 +170,28 @@ boolean isOperationComplete() {
 
   /**
    * Set the exception associated with this operation.
-   * A {@link ServerErrorCode#Blob_Deleted} or {@link ServerErrorCode#Blob_Expired} error overrides any other
-   * previously received exception.
+   * First, if current operationException is null, directly set operationException as exception;
+   * Second, if operationException exists, compare ErrorCodes of exception and existing operation Exception depending
+   * on precedence level. An ErrorCode with a smaller precedence level overrides an ErrorCode with a larger precedence
+   * level. Update the operationException if necessary.
    * @param exception the {@link RouterException} to possibly set.
    */
   void setOperationException(Exception exception) {
     if (exception instanceof RouterException) {
       RouterErrorCode routerErrorCode = ((RouterException) exception).getErrorCode();
-      if (operationException.get() == null || routerErrorCode == RouterErrorCode.BlobDeleted
-          || routerErrorCode == RouterErrorCode.BlobExpired) {
+      if (operationException.get() == null) {
         operationException.set(exception);
+      } else {
+        Integer currentOperationExceptionLevel = null;
+        if (operationException.get() instanceof RouterException) {
```

Comment:
could use ternary (unless you feel otherwise):
```java
int currentOperationExceptionLevel = operationException.get() instanceof RouterException ?
  getPrecedenceLevel(((RouterException) operationException.get()).getErrorCode()) : 
  getPrecedenceLevel(RouterErrorCode.UnexpectedInternalError);
```